### PR TITLE
Add dialog-focus rule

### PR DESCRIPTION
**Pull request by masuP9**
## What does this change?

Add dialog-focus rule.

Move focus to inside dialog or set dialog after trigger. Run the test for each trigger that has the `aria-haspopup="dialog"`.

In addition, define ComputedAccessibleNode type

## Screenshots

-

## What can I check for bug fixes?

-

## References

- [Understanding Success Criterion 2\.4\.3: Focus Order](https://www.w3.org/WAI/WCAG21/Understanding/focus-order.html)
- [SCR26: Inserting dynamic content into the Document Object Model immediately following its trigger element](https://www.w3.org/WAI/WCAG21/Techniques/client-side-script/SCR26)
- [Dialog (Modal) - WAI\-ARIA Authoring Practices 1\.1](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal)
- [blink / renderer / core / aom / computed_accessible_node.idl](https://chromium.googlesource.com/chromium/src/third_party/+/master/blink/renderer/core/aom/computed_accessible_node.idl)

### CODE REVIEW

File: packages/types/src/index.ts [hunk 1]
```diff
@@ -11,3 +11,4 @@ export * from './schema';
 export * from './stat';
 export * from './summary';
 export * from './timing';
+export * from './util';
```
Comment (masuP9) [inline packages/types/src/index.ts]: [need help] Where is the best place to define the type of ComputedAccessibleNode.

File: packages/acot-preset-wcag/docs/rules/focus-dialog.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# focus-dialog
+
+> Move focus to inside dialog or set dialog after trigger.
```
Comment (wadackel) [inline packages/acot-preset-wcag/docs/rules/focus-dialog.md]: Please use the short summary of rules in paragraphs :pray:

File: packages/acot-preset-wcag/docs/rules/focus-dialog.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
+# focus-dialog
```
Comment (wadackel) [inline packages/acot-preset-wcag/docs/rules/focus-dialog.md]: I prefer the rule name `dialog-focus` to match the naming conventions of other rules.

File: packages/acot-preset-wcag/src/rules/focus-dialog.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        const hasDialogRoleInParentByPressTab = await context.page.evaluate(
+          activeElementHasParentDialogRole,
+        );
+
+        context.debug({ hasDialogRoleInParentByPressTab });
+
+        if (!hasDialogRoleInParentByPressTab) {
+          await context.report({
+            node,
+            message: `move focus to inside dialog or set dialog after trigger`,
```
Comment (wadackel) [inline packages/acot-preset-wcag/src/rules/focus-dialog.ts]:
```suggestion
            message: `Move focus to inside dialog or set dialog after trigger.`,
```

File: packages/acot-preset-wcag/src/rules/focus-dialog.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+            return computedAXNode;
+          }
+
+          return computedAXNode.parent != undefined
+            ? findParentDialogRoleAXNode(computedAXNode.parent)
+            : null;
+        }
+
+        return findParentDialogRoleAXNode(ax)?.role === 'dialog';
+      };
```
Comment (wadackel) [inline packages/acot-preset-wcag/src/rules/focus-dialog.ts]: Unless you have a clear reason, I like to use `const` like any other function.

File: packages/types/src/index.ts (hunk 1 above)
Comment (wadackel) [reply packages/types/src/index.ts]: I want to avoid util because its role is unclear.
I think `src / browser.ts` is valid.

**Review by wadackel — CHANGES_REQUESTED** [review]

File: packages/types/src/index.ts (hunk 1 above)
Comment (masuP9) [reply packages/types/src/index.ts]: `src/browser.ts` it sounds good and is clear type role.

File: packages/acot-preset-wcag/src/rules/focus-dialog.ts (hunk 5 above)
Comment (masuP9) [reply packages/acot-preset-wcag/src/rules/focus-dialog.ts]: OK, I don't have big reason. I fixed.

**Review by wadackel — APPROVED** [review]
Thanks for implement the rule!
LGTM :+1:

**Pull request merged**